sort: avoid sigpipe errors

miDeb opened the pull request:
By calling `unwrap` we get a panic instead of an abort, and since we
mute sigpipe panics for all utilites, no error message will be printed.

tertsdiepraam commented:
I hope I'm understanding this correctly: the problem will still occur but now fail silently? So I guess this is something that we can't actually fix in the code?

miDeb commented:
My understanding is that when we're trying to write to stdout, but the pipe has already been closed, we get a "Broken Pipe" error (not always though, seems to be random). That could happen for example when you run `seq 100000 | sort | head`: Head will close the pipe after having read the first lines, and when we try to write the next time we get `Broken Pipe`.
So, yeah, this only silences that error message and we exit normally.

By replacing the `crash_if_err! ` with an `unwrap` we get a panic. Panics containing `Broken pipe` were already ignored because all utilities call `mute_sigpipe_panic` which installs a panic hook that ignores them.
Because I added a test I noticed that on Windows the panic message does not contain `Broken pipe`, so I changed that.

tertsdiepraam commented:
Right, that makes sense! Having a system to deal with the broken pipe is great. I guess it's probably not worth it to agonize over those tests or the CI to figure out why that broken pipe occurs?

miDeb commented:
Is this still occurring on CI? The reason why it occurs only sometimes is timing, and the reason why we get a `SIGPIPE` (broken pipe) is that our stdout got closed by the reader, potentially because it died. I don't believe we should show a `Broken Pipe` error, just silently exit.

tertsdiepraam commented:
Sorry, I didn't mean to imply that it still happens with this fix. Just wondering why our stdout gets closed sometimes and whether we should have another PR that would make sure that it doesn't get closed in the tests.

miDeb commented:
So does this happen for tests? I'm pretty sure tests nicely wait for the child to terminate by calling `wait_with_output`, unless you do something like I did in the test I added.

tertsdiepraam commented:
I'm confused now😄 I thought you were fixing a problem in the CI. If you're not then it's all good. I must've misremembered some broken pipe error for another util

miDeb commented:
No, I discovered this running a command locally :) the fix for a broken pipe in CI was https://github.com/uutils/coreutils/pull/2419

tertsdiepraam commented:
Ah got it! Thanks!
